File: pptpro/frontend/src/api/slideContent.ts
/**
 * Slide content API client - classification and generation
 */
import apiClient from './client';

export interface SlideClassificationRequest {
  slide_text: string;
  slide_type: string;
  head_message?: string;
}

export interface ContentElement {
  element_type: string;
  description: string;
  classification: 'USER_NEEDED' | 'AI_GENERATED';
  reason: string;
}

export interface SlideClassificationResponse {
  user_needed: ContentElement[];
  ai_generated: ContentElement[];
}

export interface SlideContentGenerationRequestContext {
  [key: string]: string | number | boolean | null | undefined;
}

export interface SlideContentGenerationRequest {
  slide_type: string;
  ai_generated_elements: string[];
  context: SlideContentGenerationRequestContext;
}

export interface MessageOnlyPayload {
  main_message?: string;
  supporting_points?: string[];
  call_to_action?: string;
}

export interface AsIsToBePayload {
  as_is_title?: string;
  as_is_points?: string[];
  to_be_title?: string;
  to_be_points?: string[];
  transition_method?: string;
}

export interface CaseBoxItem {
  title?: string;
  description?: string;
  pros?: string[];
  cons?: string[];
  recommendation?: string;
}

export interface CaseBoxPayload {
  cases?: CaseBoxItem[];
  insight_box?: string;
}

export interface StepItem {
  order?: number;
  title?: string;
  description?: string;
  deliverables?: string[];
  timeline?: string;
}

export interface StepFlowPayload {
  steps?: StepItem[];
  action_guide?: string;
}

export interface ChartInsightPayload {
  chart_title?: string;
  chart_type?: string;
  key_insights?: string[];
  data_source?: string;
  evidence_block?: string;
  insight_box?: string;
}

export interface ConnectionItem {
  from: string;
  to: string;
  relationship?: string;
}

export interface NodeMapPayload {
  central_concept?: string;
  primary_nodes?: string[];
  connections?: ConnectionItem[];
  insight_box?: string;
}

export type TemplatePayload =
  | MessageOnlyPayload
  | AsIsToBePayload
  | CaseBoxPayload
  | StepFlowPayload
  | ChartInsightPayload
  | NodeMapPayload;

export interface SlideComponents {
  title?: string;
  sub_message?: string;
  bullet_points?: string[];
  evidence_block?: string;
  diagram_components?: Record<string, unknown>;
  insight_box?: string;
  action_guide?: string;
  caption?: string;
  main_message?: string;
  supporting_points?: string[];
  call_to_action?: string;
  as_is_title?: string;
  as_is_points?: string[];
  to_be_title?: string;
  to_be_points?: string[];
  transition_method?: string;
  cases?: CaseBoxItem[];
  steps?: StepItem[];
  chart_title?: string;
  chart_type?: string;
  key_insights?: string[];
  data_source?: string;
  central_concept?: string;
  primary_nodes?: string[];
  connections?: ConnectionItem[];
  ppt_payload?: TemplatePayload;
  [key: string]: unknown;
}

export interface SlideContentGenerationResponse {
  components: SlideComponents;
  metadata?: Record<string, unknown>;
}

export const slideContentApi = {
  /**
   * 슬라이드 콘텐츠를 USER_NEEDED / AI_GENERATED로 자동 분류
   */
  classifyContent: async (
    request: SlideClassificationRequest
  ): Promise<SlideClassificationResponse> => {
    const response = await apiClient.post<SlideClassificationResponse>(
      '/api/slide/classify',
      request
    );
    return response.data;
  },

  /**
   * AI_GENERATED 요소들에 대해 실제 콘텐츠 생성
   */
  generateContent: async (
    request: SlideContentGenerationRequest
  ): Promise<SlideContentGenerationResponse> => {
    const response = await apiClient.post<SlideContentGenerationResponse>(
      '/api/slide/generate',
      request
    );
    return response.data;
  },
};

File: pptpro/backend/app/schemas/slide_content.py
"""
Slide content generation schemas
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any


class SlideClassificationRequest(BaseModel):
    """슬라이드 분류 요청 - USER_NEEDED / AI_GENERATED 구분"""
    slide_text: str = Field(..., description="분류할 슬라이드 텍스트 또는 개요")
    slide_type: str = Field(..., description="슬라이드 타입 (template_type)")
    head_message: Optional[str] = Field(None, description="헤드 메시지")


class ContentElement(BaseModel):
    """콘텐츠 요소"""
    element_type: str = Field(..., description="요소 타입 (title, bullet_point, data, insight 등)")
    description: str = Field(..., description="요소 설명")
    classification: str = Field(..., description="USER_NEEDED 또는 AI_GENERATED")
    reason: str = Field(..., description="분류 이유")


class SlideClassificationResponse(BaseModel):
    """슬라이드 분류 응답"""
    user_needed: List[ContentElement] = Field(..., description="사용자 입력이 필요한 요소들")
    ai_generated: List[ContentElement] = Field(..., description="AI가 생성 가능한 요소들")


class SlideContentGenerationRequest(BaseModel):
    """슬라이드 콘텐츠 생성 요청"""
    slide_type: str = Field(..., description="슬라이드 템플릿 타입")
    ai_generated_elements: List[str] = Field(..., description="AI가 생성할 요소 타입 리스트")
    context: Dict[str, Any] = Field(..., description="생성에 필요한 컨텍스트 (주제, 목표, 헤드메시지 등)")


class SlideComponents(BaseModel):
    """생성된 슬라이드 컴포넌트"""
    title: Optional[str] = Field(None, description="슬라이드 제목")
    sub_message: Optional[str] = Field(None, description="서브 메시지")
    bullet_points: Optional[List[str]] = Field(None, description="불릿 포인트 리스트")
    evidence_block: Optional[str] = Field(None, description="근거/증거 블록")
    diagram_components: Optional[Dict[str, Any]] = Field(None, description="다이어그램 구성 요소")
    insight_box: Optional[str] = Field(None, description="인사이트 박스")
    action_guide: Optional[str] = Field(None, description="액션 가이드")
    caption: Optional[str] = Field(None, description="캡션")


class SlideContentGenerationResponse(BaseModel):
    """슬라이드 콘텐츠 생성 응답"""
    components: SlideComponents = Field(..., description="생성된 컴포넌트들")
    metadata: Optional[Dict[str, Any]] = Field(None, description="추가 메타데이터")
